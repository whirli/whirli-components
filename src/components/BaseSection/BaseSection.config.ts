// @ts-ignore
import { PropType } from '@composition';
import * as PropHelpers from '@whirli-components/helpers/props';
import { ComponentStyles, ComponentProps } from '@whirli-components/@types/components';
import { createClassesFromPropValues } from '@whirli-components/helpers/classes';

// Local config
import config from '@whirli-local/components/BaseSection/BaseSection.config';

// Package config
import { PropKeys, PropValues } from './BaseSection.constants';
import { PropBackground } from './BaseSection.types';

// Types
import { PropSize } from './BaseSection.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
  [PropKeys.BACKGROUND]: {
    generateBreakpoints: true,
    classes: {
      default: 'background-default',
      ...createClassesFromPropValues(PropValues, PropKeys.BACKGROUND),
      ...config.styles[PropKeys.BACKGROUND]?.classes,
    } as Record<PropBackground, string>,
  },
};

export const ConfigProps: ComponentProps = {
  [PropKeys.BACKGROUND]: {
    required: false,
    type: [String, Object] as PropType<PropBackground>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.BACKGROUND),
  },
  [PropKeys.SIZE]: {
    required: false,
    type: [String, Object, Array] as PropType<PropSize>,
    default: 'default',
  },
  ...config.props,
};
