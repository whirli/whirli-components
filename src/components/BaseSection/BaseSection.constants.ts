import { PropValues as SpacerPropValues } from '../BaseSpacer/BaseSpacer.constants';

export const PropKeys = {
  BACKGROUND: 'background',
  SIZE: 'size',
};

export const PropValues = {
  [PropKeys.BACKGROUND]: ['default'] as string[],
  [PropKeys.SIZE]: SpacerPropValues.SIZE,
};
