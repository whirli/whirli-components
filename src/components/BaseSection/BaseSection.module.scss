@import '../../styles/Breakpoints.scss';

// default styles
.section {}
// breakpoint styles
$packageBreakpointStyles: (
    background-default: (),
);