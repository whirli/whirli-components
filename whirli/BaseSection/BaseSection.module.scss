@import '@whirli-components/src/components/BaseSection/BaseSection.module.scss';

// Standard styles

// Breakpoint styles
$localBreakpointStyles: ();

@include createBreakpointClasses(map-merge($packageBreakpointStyles, $localBreakpointStyles));