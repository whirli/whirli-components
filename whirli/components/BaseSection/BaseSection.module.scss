@import '@whirli-components/components/BaseSection/BaseSection.module.scss';

// Standard styles

// Breakpoint styles
$localBreakpointStyles: ();

@include createBreakpointClasses(map-merge($packageBreakpointStyles, $localBreakpointStyles));