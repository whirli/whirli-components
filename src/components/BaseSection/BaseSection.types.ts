import { PropKeys, PropValues } from './BaseSection.constants';

const propBackground = PropValues[PropKeys.BACKGROUND];

export type PropBackground = typeof propBackground[number];

export type PropSize =
  | string
  | Record<string, string>
  | [string | Record<string, string>, string | Record<string, string>];

export interface ComputedSize {
  top: PropSize;
  bottom: PropSize;
}
