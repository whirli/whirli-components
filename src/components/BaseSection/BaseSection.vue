<template>
  <section :class="classes">
    <BaseSpacer :size="size.top" />
    <div>
      <slot />
    </div>
    <BaseSpacer :size="size.bottom" />
  </section>
</template>

<script setup lang="ts">
// Vue
import { computed } from 'vue';

// Types
import type { ComputedRef } from 'vue';
import { ComputedSize } from './BaseSection.types';

// Components
import BaseSpacer from '@whirli-components/components/BaseSpacer/BaseSpacer.vue';

// Styles
import styles from '@whirli-local/components/BaseSection/BaseSection.module.scss';

// Data
import { ConfigStyles, ConfigProps } from './BaseSection.config';

const ComponentStyles = ConfigStyles;

const props = defineProps(ConfigProps);

// Classes
import useClasses from '@whirli-components/@use/class';
const { makeClasses } = useClasses();
const classes = [styles.section, ...makeClasses(ComponentStyles, props, styles)];

const size: ComputedRef<ComputedSize> = computed(() => {
  return {
    top: props.size.top ? props.size.top : props.size.bottom ? undefined : props.size,
    bottom: props.size.bottom ? props.size.bottom : props.size.top ? undefined : props.size,
  };
});
</script>
