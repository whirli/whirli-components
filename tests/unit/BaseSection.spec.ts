import { shallowMount } from '@vue/test-utils';

import BaseSection from '@/components/BaseSection/BaseSection.vue';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let wrapper: any;
const text = 'Hello world';

describe('BaseSection.vue - Single classes: Equal top and bottom', () => {
  beforeEach(() => {
    wrapper = shallowMount(BaseSection, {
      props: {
        size: 'md',
      },
      slots: {
        default: text,
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('returns the correct top spacer value', () => {
    expect(wrapper.vm.gutter.top).toBe('md');
  });

  it('returns the correct bottom spacer value', () => {
    expect(wrapper.vm.gutter.bottom).toBe('md');
  });

  it('displays the correct content', () => {
    expect(wrapper.find('div').text()).toBe(text);
  });
});

describe('BaseSection.vue - Single classes: Seperate top and bottom', () => {
  beforeEach(() => {
    wrapper = shallowMount(BaseSection, {
      props: {
        size: {
          top: 'md',
          bottom: 'lg',
        },
      },
      slots: {
        default: text,
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('returns the correct top spacer value', () => {
    expect(wrapper.vm.gutter.top).toBe('md');
  });

  it('returns the correct bottom spacer value', () => {
    expect(wrapper.vm.gutter.bottom).toBe('lg');
  });

  it('displays the correct content', () => {
    expect(wrapper.find('div').text()).toBe(text);
  });
});
