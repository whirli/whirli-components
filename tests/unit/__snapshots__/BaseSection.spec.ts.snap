// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseSection.vue - Single classes: Equal top and bottom renders the component 1`] = `
<section
  class=""
>
  <base-spacer-stub
    size="md"
  />
  <div>
    
    Hello world
    
  </div>
  <base-spacer-stub
    size="md"
  />
</section>
`;

exports[`BaseSection.vue - Single classes: Seperate top and bottom renders the component 1`] = `
<section
  class=""
>
  <base-spacer-stub
    size="md"
  />
  <div>
    
    Hello world
    
  </div>
  <base-spacer-stub
    size="lg"
  />
</section>
`;
